core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -324,6 +324,8 @@ public static RexNode splitFilter(final RexSimplify simplify,
    */
   private static RexNode canonizeNode(RexBuilder rexBuilder, RexNode condition) {
     switch (condition.getKind()) {
+    case PLUS:
```

Comment:
The code in this block removes duplicate operands (via the `TreeMap`). You will need slightly different code for `PLUS` or `TIMES` (just the sort piece).

Once this is fixed, could you add a test with duplicate operand, e.g,., `x + x` ?

Reply:
@jcamachor Thanks for catching this. I have updated the pull request.